File: src/server/CloudServerManager.php
<?php

namespace pocketcloud\cloud\server;

use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\group\ServerGroup;
use pocketcloud\cloud\network\packet\impl\ServerSyncPacket;
use pocketcloud\cloud\server\data\CloudServerData;
use pocketcloud\cloud\server\util\ServerStartMethod;
use pocketcloud\cloud\server\util\ServerStatus;
use pocketcloud\cloud\server\util\ServerUtils;
use pocketcloud\cloud\template\Template;
use pocketcloud\cloud\template\TemplateType;
use pocketcloud\cloud\util\benchmark\Benchmark;
use pocketcloud\cloud\util\misc\Queue;
use pocketcloud\cloud\util\misc\Tickable;
use pocketcloud\cloud\util\promise\Promise;
use pocketcloud\cloud\util\trait\SingletonTrait;
use Ramsey\Uuid\Uuid;
use Throwable;

final class CloudServerManager implements Tickable {
    use SingletonTrait;

    private const int MULTI_START_THRESHOLD = 5;
    private const int MULTI_START_BATCH_SIZE = 10;

    /** @var array<CloudServer> */
    private array $servers = [];
    private float $lastServerStartTime = 0;
    private float $lastServerStopTime = 0;

    private float $nextServerStartTime = 0;

    /** @var array<string> */
    private array $latestServerStartTimes = [];

    /** @var Queue<CloudServer> */
    private Queue $serverPrepareQueue;
    /** @var Queue<CloudServer> */
    private Queue $serverStartQueue;

    public function __construct() {
        self::setInstance($this);
        $this->serverPrepareQueue = Queue::fromClass(CloudServer::class);
        $this->serverStartQueue = Queue::fromClass(CloudServer::class);
    }

    public function start(Template $template, int $count = 1): array {
        $startedServers = [];
        if (!$this->checkCapacity($template)) {
            CloudLogger::get()->warn("Failed to start any more servers of §b{} §rdue to the max amount of servers already being reached.", $template->getName());
        } else {
            for ($i = 0; $i < $count; $i++) {
                if (!$this->checkCapacity($template)) break;
                $this->lastServerStartTime = microtime(true);
                $id = ServerUtils::getFreeId($template);
                if ($id !== -1) {
                    $port = ServerUtils::getFreePort($template->getTemplateType());
                    if ($port > 0) {
                        $server = new CloudServer($id, Uuid::uuid4()->toString(), $template->getName(), new CloudServerData($template->getName() . "-" . $id, $port, $template->getSettings()->getMaxPlayerCount(), null));
                        $this->latestServerStartTimes[$template->getName()] = $server->getName();
                        $this->add($server);
                        $this->serverPrepareQueue->add($server);
                        $startedServers[] = $server->getName();
                    } else {
                        CloudLogger::get()->warn("Failed to start any more servers of §b{}§8: §cNo available ports found.", $template->getName());
                        break;
                    }
                }
            }
        }

        return $startedServers;
    }

    public function save(CloudServer $server): Promise {
        $promise = new Promise();
        $saveCommandLine = $server->getTemplate()->getTemplateType()->getSaveCommandLine();
        if ($saveCommandLine === null) {
            $server->saveFiles();
            return Promise::resolved();
        }

        $server->executeCommand($saveCommandLine)->then(function () use ($server, $promise) {
            $server->saveFiles();
            $promise->resolve();
        })->failure(fn() => $promise->reject("Request timeout"));

        return $promise;
    }

    public function stop(CloudServer|Template|ServerGroup|string $source, bool $force): array {
        if ($source instanceof CloudServer) {
            $affectedServers = [$source];
        } else {
            if (is_string($source) && ($server = $this->get($source)) !== null) {
                $affectedServers = [$server];
            } else {
                $affectedServers = $this->getAll($source);
            }
        }

        foreach ($affectedServers as $server) $server->stop($force);
        return $affectedServers;
    }

    public function stopAll(bool $force = false): array {
        foreach (($servers = $this->getAll()) as $server) $server->stop($force);
        return $servers;
    }

    public function add(CloudServer $server): void {
        if (!isset($this->servers[$server->getName()])) $this->servers[$server->getName()] = $server;
        ServerUtils::addId($server->getTemplate(), $server->getId());
        ServerUtils::addPort($server->getServerData()->getPort());
    }

    public function remove(CloudServer $server): void {
        if (isset($this->servers[$server->getName()])) unset($this->servers[$server->getName()]);
        ServerUtils::removeId($server->getTemplate(), $server->getId());
        ServerUtils::removePort($server->getServerData()->getPort());
        $this->lastServerStopTime = microtime(true);
        ServerSyncPacket::create($server, true)->broadcastPacket();
    }

    public function checkCapacity(Template $template): bool {
        return count($this->getAll($template)) < $template->getSettings()->getMaxServerCount();
    }

    private function addToStartQueue(CloudServer $server): void {
        CloudLogger::get()->debug("Done preparing server: §b{}", $server->getName());
        $this->serverStartQueue->add($server);
    }

    private function onStartFailed(array $crashData): void {
        [$server, $exception] = $crashData;
        CloudLogger::get()->warn("§cFailed to prepare server §e{}§8: §e", $server, $exception?->getMessage() ?? "Unknown error");
        if ($exception instanceof Throwable) CloudLogger::get()->exception($exception);
    }

    public function tick(int $currentTick): void {
        //keep alive, timeout,... etc
        // ABOVE > QUEUES (first the above, then the queues, SERVER BY SERVER (1 server/tick).
        foreach ($this->servers as $server) $server->tick($currentTick);

        if (!$this->serverPrepareQueue->isEmpty()) {
            Benchmark::startTiming("check_server_prepare_queue");
            $this->serverPrepareQueue->next()->prepare()
                ->then($this->addToStartQueue(...))
                ->failure($this->onStartFailed(...));
            Benchmark::stopTiming("check_server_prepare_queue");
            return;
        }

        Benchmark::startTiming("check_server_start_queue");
        if ($currentTick >= $this->nextServerStartTime && !$this->serverStartQueue->isEmpty()) {
            $method = ServerStartMethod::current();
            $queueSize = $this->serverStartQueue->count();

            if ($method->supportsMultiStart() && $queueSize >= self::MULTI_START_THRESHOLD) {
                /** @var array<CloudServer> $batch */
                $batch = [];
                $limit = min($queueSize, self::MULTI_START_BATCH_SIZE);
                for ($i = 0; $i < $limit; $i++) {
                    $batch[] = $this->serverStartQueue->next();
                }

                foreach ($batch as $server) {
                    $server->start(false);
                }

                if ($method->multiStartServer($batch)) {
                    foreach ($batch as $server) $server->handleStartSuccess(null);
                } else {
                    foreach ($batch as $server) $server->handleFailedStart(true);
                }
            } else {
                $this->serverStartQueue->next()->start();
            }

            $this->nextServerStartTime = $currentTick + 10;
        }

        Benchmark::stopTiming("check_server_start_queue");
    }

    public function get(string $name): ?CloudServer {
        return $this->servers[$name] ?? array_find($this->servers, fn(CloudServer $server) => $server->getServerUuid() == $name);
    }

    public function getLatest(Template $template): ?CloudServer {
        if (!isset($this->latestServerStartTimes[$template->getName()])) return null;
        return $this->servers[$this->latestServerStartTimes[$template->getName()]] ?? null;
    }

    public function getAll(Template|TemplateType|ServerGroup|string|null ...$templateOrGroups): array {
        if (count($templateOrGroups) > 0) return array_filter($this->servers, function (CloudServer $server) use ($templateOrGroups): bool {
            foreach ($templateOrGroups as $templateOrGroup) {
                if ($templateOrGroup === null) continue;
                $templateOrGroup = is_string($templateOrGroup) ? $templateOrGroup : $templateOrGroup->getName();
                return $server->getName() == $templateOrGroup ||
                    $server->getTemplateName() == $templateOrGroup ||
                    $server->getTemplate()?->getParentServerGroup()?->getName() == $templateOrGroup ||
                    $server->getTemplate()?->getTemplateType()->getName() == $templateOrGroup;
            }

            return true;
        });

        return $this->servers;
    }

    public function getLastServerStartTime(): float {
        return $this->lastServerStartTime;
    }

    public function getLastServerStopTime(): float {
        return $this->lastServerStopTime;
    }
}

File: src/main/java/de/pocketcloud/cloud/server/CloudServerManager.java
package de.pocketcloud.cloud.server;

import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.network.packet.impl.ServerSyncPacket;
import de.pocketcloud.cloud.server.util.CloudServerData;
import de.pocketcloud.cloud.server.start.ServerStartMethod;
import de.pocketcloud.cloud.server.util.ServerStartMethods;
import de.pocketcloud.cloud.server.util.ServerUtils;
import de.pocketcloud.cloud.template.Template;
import de.pocketcloud.cloud.template.TemplateType;
import de.pocketcloud.cloud.template.group.ServerGroup;
import de.pocketcloud.cloud.tick.Tickable;
import de.pocketcloud.cloud.util.benchmark.Benchmark;
import de.pocketcloud.cloud.util.concurrent.Promise;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Getter
@Accessors(fluent = true)
public final class CloudServerManager implements Tickable {

    public static final ExecutorService SERVER_EXECUTOR = Executors.newFixedThreadPool(
            Runtime.getRuntime().availableProcessors(),
            r -> {
                Thread t = new Thread(r, "Server-Worker");
                t.setDaemon(false);
                return t;
            }
    );

    @Getter
    @Accessors(fluent = true)
    private static CloudServerManager instance = null;

    private static final int MULTI_START_THRESHOLD = 5;
    private static final int MULTI_START_BATCH_SIZE = 10;

    @Getter(AccessLevel.NONE)
    private final Map<String, CloudServer> servers = new ConcurrentHashMap<>();
    private long lastServerStartTime = 0;
    private long lastServerStopTime = 0;
    private long nextServerStartTime = 0;

    private final Map<String, String> latestServerStartTimes = new ConcurrentHashMap<>();

    @Getter(AccessLevel.NONE)
    private final Queue<CloudServer> serverPrepareQueue = new LinkedList<>();
    @Getter(AccessLevel.NONE)
    private final Queue<CloudServer> serverStartQueue = new LinkedList<>();

    public CloudServerManager() {
        instance = this;
    }

    public List<String> start(Template template, int count) {
        List<String> startedServers = new ArrayList<>();

        if (!checkCapacity(template)) {
            CloudLogger.get().warn("Failed to start any more servers of §b{} §rdue to the max amount of servers already being reached.", template.name());
            return startedServers;
        }

        for (int i = 0; i < count; i++) {
            if (!checkCapacity(template)) break;
            this.lastServerStartTime = System.currentTimeMillis();

            int id = ServerUtils.getFreeId(template);
            if (id == -1) continue;

            int port = ServerUtils.getFreePort(template.templateType());
            if (port <= 0) {
                CloudLogger.get().warn("Failed to start any more servers of §b{}§8: §cNo available ports found.", template.name());
                break;
            }

            UUID uuid = UUID.randomUUID();
            CloudServer server = new CloudServer(
                id,
                uuid,
                template.name(),
                new CloudServerData(uuid, port, template.settings().getMaxPlayerCount())
            );

            latestServerStartTimes.put(template.name(), server.name());
            add(server);
            serverPrepareQueue.offer(server);
            startedServers.add(server.name());
        }

        return startedServers;
    }

    public Promise<Void> save(CloudServer server) {
        String saveCommandLine = server.template().serverSoftware().config().saveCommandLine();
        if (saveCommandLine == null) {
            return server.save();
        }

        Promise<Void> promise = new Promise<>();
//        server.executeCommand(saveCommandLine)
//            .then(() -> {
//                server.saveFiles();
//                promise.resolve();
//            })
//            .failure(() -> promise.reject("Request timeout"));

        return promise;
    }

    public List<CloudServer> stop(CloudServer source, boolean force) {
        List<CloudServer> affectedServers = new ArrayList<>(Collections.singleton(source));
        affectedServers.forEach(server -> server.stop(force));
        return affectedServers;
    }

    public List<CloudServer> stop(Template template, boolean force) {
        List<CloudServer> affectedServers = getAll(template);
        affectedServers.forEach(server -> server.stop(force));
        return affectedServers;
    }

    public List<CloudServer> stop(ServerGroup group, boolean force) {
        List<CloudServer> affectedServers = getAll(group);
        affectedServers.forEach(server -> server.stop(force));
        return affectedServers;
    }

    public List<CloudServer> stop(TemplateType templateType, boolean force) {
        List<CloudServer> affectedServers = getAll(templateType);
        affectedServers.forEach(server -> server.stop(force));
        return affectedServers;
    }

    public List<CloudServer> stop(String name, boolean force) {
        List<CloudServer> affectedServers = new ArrayList<>();
        if (servers.containsKey(name)) affectedServers.add(servers.get(name));
        affectedServers.forEach(server -> server.stop(force));
        return affectedServers;
    }

    public List<CloudServer> stopAll(boolean force) {
        List<CloudServer> all = getAll();
        all.forEach(server -> server.stop(force));
        return all;
    }

    public void add(CloudServer server) {
        servers.putIfAbsent(server.name(), server);
        ServerUtils.addId(server.template(), server.id());
        ServerUtils.addPort(server.serverData().port());
    }

    public void remove(CloudServer server) {
        servers.remove(server.name());
        ServerUtils.removeId(server.template(), server.id());
        ServerUtils.removePort(server.serverData().port());
        this.lastServerStopTime = System.currentTimeMillis();
        ServerSyncPacket.create(server, true).broadcastPacket();
    }

    public boolean checkCapacity(Template template) {
        return getAll(template).size() < template.settings().getMaxServerCount();
    }

    private void addToStartQueue(CloudServer server) {
        CloudLogger.get().debug("Done preparing server: §b{}", server.name());
        serverStartQueue.offer(server);
    }

    private void onStartFailed(Object[] crashData) {
        CloudServer server = (CloudServer) crashData[0];
        Throwable exception = crashData.length > 1 ? (Throwable) crashData[1] : null;
        CloudLogger.get().warn("§cFailed to prepare server §e{}§8: §e{}", server, exception != null ? exception.getMessage() : "Unknown error");
        if (exception != null) CloudLogger.get().exception(exception);
    }

    @Override
    public void tick(long currentTick) {
        servers.values().forEach(server -> server.tick(currentTick));

        if (!serverPrepareQueue.isEmpty()) {
            Benchmark.startTiming("check_server_prepare_queue");

            CloudServer server = serverPrepareQueue.poll();
            server.prepare()
                .thenSuccess(_ -> addToStartQueue(server))
                .failure(e -> onStartFailed(new Object[]{server, e}));

            Benchmark.stopTiming("check_server_prepare_queue");
            return;
        }

        Benchmark.startTiming("check_server_start_queue");
        if (currentTick >= nextServerStartTime && !serverStartQueue.isEmpty()) {
            ServerStartMethod method = ServerStartMethods.current();
            int queueSize = serverStartQueue.size();
            if (queueSize >= MULTI_START_THRESHOLD) {
                List<CloudServer> batch = new ArrayList<>();
                int limit = Math.min(queueSize, MULTI_START_BATCH_SIZE);
                for (int i = 0; i < limit; i++) batch.add(serverStartQueue.poll());

                batch.forEach(CloudServer::start);
                method.start(batch.toArray(new CloudServer[0])).thenSuccess(map -> {
                    for (CloudServer server : batch) {
                        if (map.containsKey(server.name())) {
                            CloudServersHandler.handleStartSuccess(server, map.get(server.name()));
                        } else {
                            CloudServersHandler.handleStartFailure(server, null, true);
                        }
                    }
                }).failure(ex -> {
                    CloudLogger.get().exception("Failed to start servers", ex);
                    for (CloudServer server : batch) {
                        CloudServersHandler.handleStartFailure(server, ex, true);
                    }
                });
            } else {
                Objects.requireNonNull(serverStartQueue.poll()).start().boot();
            }

            nextServerStartTime = currentTick + 10;
        }

        Benchmark.stopTiming("check_server_start_queue");
    }

    public Optional<CloudServer> get(String name) {
        return Optional.ofNullable(servers.getOrDefault(name,
            servers.values().stream()
                .filter(s -> s.name().startsWith(name))
                .findFirst().orElse(null)
        ));
    }

    public Optional<CloudServer> get(UUID uuid) {
        return servers.values().stream().filter(s -> s.uuid().compareTo(uuid) == 0).findFirst();
    }

    public Optional<CloudServer> getFreeLobby() {
        return servers.values().stream()
                .filter(s -> s.template().settings().isLobby() && s.playerCount() < s.serverData().maxPlayers())
                .min(Comparator.comparingInt(CloudServer::playerCount));
    }

    public Optional<CloudServer> getLatest(Template template) {
        String latestName = latestServerStartTimes.get(template.name());
        if (latestName == null) return Optional.empty();
        return Optional.ofNullable(servers.getOrDefault(latestName, null));
    }

    public List<CloudServer> getAll() {
        return new ArrayList<>(servers.values());
    }

    public List<CloudServer> getAll(Template template) {
        return servers.values().stream().filter(s -> s.templateName().equals(template.name())).toList();
    }

    public List<CloudServer> getAll(TemplateType type) {
        return servers.values().stream().filter(s -> s.template().templateType().name().equals(type.name())).toList();
    }

    public List<CloudServer> getAll(ServerGroup group) {
        return servers.values().stream().filter(s -> s.template().parentGroups().stream().anyMatch(p -> p.name().equals(group.name()))).toList();
    }
}